pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java

```diff
@@ -514,4 +529,47 @@ public List<SegmentConsumerInfo> getConsumingSegmentsInfo(
     }
     return segmentConsumerInfoList;
   }
+
+  @GET
+  @Path("tables/{tableNameWithType}/validate")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Validates if the ideal state matches with the segmentstate on this server", notes =
+      "Validates if the ideal state matches with the segmentstate on this server")
+  public TableSegmentValidationInfo validateTableIdealState(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableNameWithType")
+          String tableNameWithType) {
+    // Get table current ideal state
+    IdealState tableIdealState = HelixHelper.getTableIdealState(_helixManager, tableNameWithType);
+    TableDataManager tableDataManager =
+        ServerResourceUtils.checkGetTableDataManager(_serverInstance, tableNameWithType);
+
+    // Validate segments in idealstate which belong to this server
+    long maxEndTime = -1;
+    Map<String, Map<String, String>> instanceStatesMap = tableIdealState.getRecord().getMapFields();
+    for (Map.Entry<String, Map<String, String>> kv : instanceStatesMap.entrySet()) {
+      String segmentName = kv.getKey();
+      if (kv.getValue().containsKey(_instanceId)) {
```

Comment:
We want to check the state of the server in IS. For `ONLINE`, check the CRC; For `CONSUMING`, check if the segment exist; Skip other states.

Reply:
Ack